tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AWS/SQS/ContextPropagation.cs

```diff
@@ -84,5 +89,62 @@ public void Set(StringBuilder carrier, string key, string value)
                 carrier.AppendFormat("\"{0}\":\"{1}\",", key, value);
             }
         }
+
+        /// <summary>
+        /// The adapter to use to append stuff to a string builder where a json is being built
+        /// </summary>
+        private readonly struct StringBuilderJsonAdapter : IBinaryHeadersCollection
+        {
+            private readonly StringBuilder _carrier;
+
+            public StringBuilderJsonAdapter(StringBuilder carrier)
+            {
+                _carrier = carrier;
+            }
+
+            public byte[] TryGetLastBytes(string name)
+            {
+                throw new NotImplementedException("this adapter can only be use to write to a StringBuilder, not to read data");
+            }
+
+            public void Add(string key, byte[] value)
+            {
+                _carrier.AppendFormat("\"{0}\":\"{1}\",", key, Convert.ToBase64String(value));
+            }
+        }
+
+        /// <summary>
+        /// The adapter to use to read attributes packed in a json string under the _datadog key
+        /// </summary>
+        public readonly struct MessageAttributesAdapter : IBinaryHeadersCollection
+        {
+            private readonly IDictionary _messageAttributes;
+
+            public MessageAttributesAdapter(IDictionary messageAttributes)
+            {
+                _messageAttributes = messageAttributes;
+            }
+
+            public byte[] TryGetLastBytes(string name)
```

Comment:
This is only being called once right? If so, that's probably fine: if not, we'll be repeatedly ducktyping and deserializing the attributes, so would be better to do that in the constructor. If we _know_ we're always going to call this method _at least_ once, then it probably still makes sense to do that work in the constructor, just in case?